mikrotik/resource_scheduler.go

```diff
@@ -2,144 +2,219 @@ package mikrotik
-	return &schema.Resource{
-		Description: "Creates a Mikrotik scheduler.",
+type scheduler struct {
+	client *client.Mikrotik
+}
 
-		CreateContext: resourceSchedulerCreate,
-		ReadContext:   resourceSchedulerRead,
-		UpdateContext: resourceSchedulerUpdate,
-		DeleteContext: resourceSchedulerDelete,
-		Importer: &schema.ResourceImporter{
-			StateContext: schema.ImportStatePassthroughContext,
-		},
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource                = &scheduler{}
+	_ resource.ResourceWithConfigure   = &scheduler{}
+	_ resource.ResourceWithImportState = &scheduler{}
+)
+
+// NewSchedulerResource is a helper function to simplify the provider implementation.
+func NewSchedulerResource() resource.Resource {
+	return &scheduler{}
+}
 
-		Schema: map[string]*schema.Schema{
-			"name": {
-				Type:        schema.TypeString,
+func (s *scheduler) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	s.client = req.ProviderData.(*client.Mikrotik)
+}
+
+// Metadata returns the resource type name.
+func (s *scheduler) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_scheduler"
+}
+
+// Schema defines the schema for the resource.
+func (s *scheduler) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "Creates a Mikrotik scheduler.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+				Description: "Identifier of this resource assigned by RouterOS",
+			},
+			"name": schema.StringAttribute{
 				Required:    true,
 				Description: "Name of the task.",
 			},
-			"on_event": {
-				Type:        schema.TypeString,
+			"on_event": schema.StringAttribute{
 				Required:    true,
 				Description: "Name of the script to execute. It must exist `/system script`.",
 			},
-			"start_date": {
-				Type:        schema.TypeString,
-				Computed:    true,
+			"start_date": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
 				Description: "Date of the first script execution.",
 			},
-			"start_time": {
-				Type:        schema.TypeString,
-				Computed:    true,
+			"start_time": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
 				Description: "Time of the first script execution.",
 			},
-			"interval": {
-				Type:        schema.TypeInt,
+			"interval": schema.Int64Attribute{
 				Optional:    true,
-				Default:     0,
+				Computed:    true,
```

Comment:
Did you mean to change this? From the mikrotik docs it seems like this shouldn't be a computed property.

<img width="1404" alt="Screen Shot 2023-02-12 at 10 01 28 PM" src="https://user-images.githubusercontent.com/5855593/218382407-079f88a7-199d-4d6f-88dd-c7bbcfe682de.png">

Reply:
actually, no.
Intention was to remove `Default` only.

Reply:
now I remember why it is there: Terraform considers this attribute as `null` but RouterOS explicitly sets it to `0` (zero), and one might end up with error 'attribute was null but now 0`

Reply:
I believe setting this to computed will break existing use cases. I personally use this field to run a script every 10 mins. If it's computed, then it's not possible to supply a value to it anymore (pictured below).

<img width="371" alt="Screen Shot 2023-02-13 at 12 49 25 PM" src="https://user-images.githubusercontent.com/5855593/218571788-88c08bf9-fb59-411b-8e94-36f396f2a00f.png">


So it seems like we need to handle that `attribute was null but now 0` error another way. Why is the `default` field problematic? The mikrotik API defaults to that value and so I believe that is the source of the "was null but now 0" error.

Reply:
that's pure Terraform thingy.
Regarding 
>I believe setting this to computed will break existing use cases

I didn't get that 🤷 
If the attribute is `Computed` + `Optional` - it means you can pass a value, but in case of `null` this value may be computed by the remote `API` server.

The case you described (when this value cannot be set) appears when you have `Computed: true` and `Optional: false` - in that case, the attribute is considered as `read-only`

Just checked locally and I am able to pass new value to `interval`.

Reply:
But if you think using `default` is cleaner approach, I can take a look (I didn't see ready solution in `plugin-framework` so I might need to implement it as new `planmodifier`)

Reply:
It's been a long time since I've looked at how these settings get composed. As long as the use case I mentioned worked, then this is good with me 👍.